### PR TITLE
Check AtomicReference's value by Predicate and Consumer

### PR DESCRIPTION
**Pull request by valery1707**
#### Check List:
* There are no opened issues
* Unit tests: YES
* Javadoc with a code example (on API only): YES

### CODE REVIEW

**Comment by valery1707** [comment]
Do I need to add `@since` tag to JavaDoc? With which value?

**Comment by valery1707** [comment]
@joel-costigliola ping

**Comment by valery1707** [comment]
@joel-costigliola ping

**Comment by joel-costigliola** [comment]
@valery1707 thanks we'll look into it for 3.18.0

File: src/main/java/org/assertj/core/api/AtomicReferenceAssert.java [hunk 1]
```diff
@@ -74,4 +81,97 @@ public AtomicReferenceAssert(AtomicReference<V> actual) {
...
+    return hasValueMatching(predicate, PredicateDescription.GIVEN);
+  }
+
+  /**
+   * Verifies that the actual atomic has the value that matches with the given predicate,
+   * the String parameter is used in the error message.
+   * <p>
+   * Example:
+   * <pre><code class='java'>
+   * // assertion will pass
```
Comment (joel-costigliola) [inline src/main/java/org/assertj/core/api/AtomicReferenceAssert.java:113]: join lines and leave one space after `<pre><code class='java'>`:
```
<pre><code class='java'> // assertion will pass
```

File: src/main/java/org/assertj/core/api/AtomicReferenceAssert.java [hunk 2]
```diff
@@ -74,4 +81,97 @@ public AtomicReferenceAssert(AtomicReference<V> actual) {
...
+   * assertThat(new AtomicReference("foo")).hasValueMatching(result -&gt; result != null, "expected not null");
+   *
+   * // assertion will fail
+   * assertThat(new AtomicReference("foo")).hasValueMatching(result -&gt; result == null, "expected null");
+   * </code></pre>
```
Comment (joel-costigliola) [inline src/main/java/org/assertj/core/api/AtomicReferenceAssert.java:118]: join lines

File: src/main/java/org/assertj/core/api/AtomicReferenceAssert.java [hunk 3]
```diff
@@ -74,4 +81,97 @@ public AtomicReferenceAssert(AtomicReference<V> actual) {
...
+   * @param requirements to assert on the actual object - must not be null.
+   * @return this assertion object.
+   *
+   * @throws NullPointerException if given Consumer is null
+   * @throws AssertionError if the actual atomic is {@code null}.
+   * @throws AssertionError if the actual atomic value does not satisfies with the given requirements.
+   *
+   * @since 3.18.0
+   */
+  public AtomicReferenceAssert<V> hasValueSatisfies(Consumer<? super V> requirements) {
```
Comment (joel-costigliola) [inline src/main/java/org/assertj/core/api/AtomicReferenceAssert.java:166]: rename to `hasValueSatisfying`

File: src/main/java/org/assertj/core/api/AtomicReferenceAssert.java [hunk 4]
```diff
@@ -74,4 +81,97 @@ public AtomicReferenceAssert(AtomicReference<V> actual) {
...
+   * @param predicate   the {@link Predicate} to apply on the resulting value.
+   * @param description the {@link Predicate} description.
+   * @return {@code this} assertion object.
+   * @throws AssertionError if the actual atomic is {@code null}.
+   * @throws AssertionError if the actual atomic value does not matches with the given predicate.
+   *
+   * @since 3.18.0
+   */
+  public AtomicReferenceAssert<V> hasValueMatching(Predicate<? super V> predicate, String description) {
+    return hasValueMatching(predicate, new PredicateDescription(description));
```
Comment (joel-costigliola) [inline src/main/java/org/assertj/core/api/AtomicReferenceAssert.java:129]: nice!

File: src/test/java/org/assertj/core/api/atomic/reference/AtomicReferenceAssert_hasValueMatching_Test.java [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+import java.util.concurrent.atomic.AtomicReference;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatExceptionOfType;
+import static org.assertj.core.util.FailureMessages.actualIsNull;
+
+class AtomicReferenceAssert_hasValueMatching_Test {
+
+  @Test
+  void should_pass_when_actual_has_value_which_is_match() {
```
Comment (joel-costigliola) [inline src/test/java/org/assertj/core/api/atomic/reference/AtomicReferenceAssert_hasValueMatching_Test.java:26]: rename to `should_pass_when_actual_has_value_matching_predicate`

File: src/test/java/org/assertj/core/api/atomic/reference/AtomicReferenceAssert_hasValueMatching_Test.java [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+    String initialValue = "FOO";
+    String expectedValue = "foo";
+
+    AtomicReference<String> actual = new AtomicReference<>(initialValue);
+
+    assertThat(actual).hasValueMatching(expectedValue::equalsIgnoreCase);
+  }
+
+  @Test
+  void should_fail_when_actual_has_value_which_does_not_match_without_description() {
```
Comment (joel-costigliola) [inline src/test/java/org/assertj/core/api/atomic/reference/AtomicReferenceAssert_hasValueMatching_Test.java:36]: rename: `should_pass_when_actual_has_value_not_matching_predicate_without_description`

File: src/test/java/org/assertj/core/api/atomic/reference/AtomicReferenceAssert_hasValueMatching_Test.java [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+
+    assertThatExceptionOfType(AssertionError.class)
+      .isThrownBy(() -> assertThat(actual).hasValueMatching(expectedValue::equalsIgnoreCase))
+      .withMessageContaining("<\"" + initialValue + "\">")
+      .withMessageContaining("to match given predicate")
+      .withMessageContaining("a better error message");
+  }
+
+  @Test
+  void should_fail_when_actual_has_value_which_does_not_match_with_description() {
```
Comment (joel-costigliola) [inline src/test/java/org/assertj/core/api/atomic/reference/AtomicReferenceAssert_hasValueMatching_Test.java:50]: rename: `should_pass_when_actual_has_value_not_matching_predicate_with_description`

File: src/test/java/org/assertj/core/api/atomic/reference/AtomicReferenceAssert_hasValueMatching_Test.java [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+    String initialValue = "foo";
+    String expectedValue = "bar";
+
+    AtomicReference<String> actual = new AtomicReference<>(initialValue);
+
+    assertThatExceptionOfType(AssertionError.class)
+      .isThrownBy(() -> assertThat(actual).hasValueMatching(expectedValue::equalsIgnoreCase))
+      .withMessageContaining("<\"" + initialValue + "\">")
+      .withMessageContaining("to match given predicate")
+      .withMessageContaining("a better error message");
```
Comment (joel-costigliola) [inline src/test/java/org/assertj/core/api/atomic/reference/AtomicReferenceAssert_hasValueMatching_Test.java:46]: use [`withMessageContainingAll`](https://www.javadoc.io/static/org.assertj/assertj-core/3.17.2/org/assertj/core/api/ThrowableAssertAlternative.html#withMessageContainingAll(java.lang.CharSequence...)) assertion

File: src/test/java/org/assertj/core/api/atomic/reference/AtomicReferenceAssert_hasValueSatisfies_Test.java [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+
+  @Test
+  void should_fail_when_requirements_is_null() {
+    String initialValue = "foo";
+
+    AtomicReference<String> actual = new AtomicReference<>(initialValue);
+
+    assertThatExceptionOfType(NullPointerException.class)
+      .isThrownBy(() -> assertThat(actual).hasValueSatisfies(null))
+      .withMessageContaining("Consumer");
```
Comment (joel-costigliola) [inline src/test/java/org/assertj/core/api/atomic/reference/AtomicReferenceAssert_hasValueSatisfies_Test.java:56]: put the full message

**Review by joel-costigliola — CHANGES_REQUESTED** [review]
PR commented, thanks @valery1707

**Pull request closed without merging**